src/coreclr/jit/fgprofile.cpp

```diff
@@ -439,443 +447,1176 @@ void BlockCountInstrumentor::InstrumentMethodEntry(Schema& schema, BYTE* profile
+//   selectively instrument; this would make merging and importing
+//   of data simpler, as all schemas for a method would agree, no
+//   matter what importer-level opts were applied.
+//
+void EfficientEdgeCountInstrumentor::Prepare(bool preImport)
 {
-    if ((block->bbFlags & BBF_HAS_CLASS_PROFILE) == 0)
+    if (!preImport)
     {
+        // If we saw badcode in the preimport prepare, we would expect
+        // compilation to blow up in the importer. So if we end up back
+        // here postimport with badcode set, something is wrong.
+        //
+        assert(!m_badcode);
         return;
     }
 
-    // Would be nice to avoid having to search here by tracking
-    // candidates more directly.
-    //
-    JITDUMP("Scanning for calls to profile in " FMT_BB "\n", block->bbNum);
-
-    // Scan the statements and add class probes
-    //
-    int classSchemaIndex = block->bbClassSchemaIndex;
-    assert((classSchemaIndex >= 0) && (classSchemaIndex < (int)schema.size()));
-
-    ClassProbeInserter                    insertProbes(schema, profileMemory, &classSchemaIndex, m_instrCount);
-    ClassProbeVisitor<ClassProbeInserter> visitor(m_comp, insertProbes);
-    for (Statement* stmt : block->Statements())
-    {
-        visitor.WalkTree(stmt->GetRootNodePointer(), nullptr);
-    }
+    JITDUMP("\nEfficientEdgeCountInstrumentor: preparing for instrumentation\n");
+    m_comp->WalkSpanningTree(this);
+    JITDUMP("%u blocks, %u probes (%u on critical edges)\n", m_blockCount, m_probeCount, m_edgeProbeCount);
 }
 
 //------------------------------------------------------------------------
-// ClassProbeInstrumentor::SuppressProbes: clean up if we're not instrumenting
+// EfficientEdgeCountInstrumentor:BuildSchemaElements: create schema
+//   elements for the probes
 //
-// Notes:
-//   Currently we're hijacking the gtCallStubAddr of the call node to hold
-//   a pointer to the profile candidate info.
+// Arguments:
+//   block -- block to instrument
+//   schema -- schema that we're building
 //
-//   We must undo this, if not instrumenting.
+// Todo: if required to have special entry probe, we must also
+//  instrument method entry with a block count.
 //
-void ClassProbeInstrumentor::SuppressProbes()
+void EfficientEdgeCountInstrumentor::BuildSchemaElements(BasicBlock* block, Schema& schema)
 {
-    unsigned                                 cleanupCount = 0;
-    SuppressProbesFunctor                    suppressProbes(cleanupCount);
-    ClassProbeVisitor<SuppressProbesFunctor> visitor(m_comp, suppressProbes);
-
-    for (BasicBlock* block = m_comp->fgFirstBB; (block != nullptr); block = block->bbNext)
+    // Walk the bbProbeList, emitting one schema element per...
+    //
+    for (Probe* probe = (Probe*)block->bbProbeList; probe != nullptr; probe = probe->next)
     {
-        if ((block->bbFlags & BBF_HAS_CLASS_PROFILE) == 0)
+        // Probe is for the edge from block->target.
```

Comment:
nit: I read `block->target` as a pointer dereference, which is confusing

Reply:
Sure, will fix.